src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -0,0 +1,595 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Xml.XPath;
+
+using Mono.Cecil;
+
+namespace Mono.Linker.Dataflow
+{
+	class XmlFlowAnnotationSource : IFlowAnnotationSource
+	{
+		readonly Dictionary<MethodDefinition, AnnotatedMethod> _methods = new Dictionary<MethodDefinition, AnnotatedMethod> ();
+		readonly Dictionary<PropertyDefinition, DynamicallyAccessedMemberTypes> _properties = new Dictionary<PropertyDefinition, DynamicallyAccessedMemberTypes> ();
+		readonly Dictionary<FieldDefinition, DynamicallyAccessedMemberTypes> _fields = new Dictionary<FieldDefinition, DynamicallyAccessedMemberTypes> ();
+
+		readonly XPathDocument _document;
+		readonly string _xmlDocumentLocation;
+		readonly LinkContext _context;
+
+		public class XmlResolutionException : Exception
+		{
+			public XmlResolutionException (string message, Exception innerException)
+				: base (message, innerException)
+			{
+			}
+		}
+
+		public XmlFlowAnnotationSource (LinkContext context, string document)
+		{
+			_xmlDocumentLocation = document;
+			_document = new XPathDocument (_xmlDocumentLocation);
+			_context = context;
+			Initialize ();
```

Comment:
This is not a great pattern, the constructor should not do such expensive operation